apps/prairielearn/src/tests/groupExam.test.ts

```diff
@@ -0,0 +1,513 @@
+  const elemList = $('form input[name="__csrf_token"]');
+  assert.lengthOf(elemList, numCsrfTokens);
+  assert.nestedProperty(elemList[0], 'attribs.value');
+  assert.isString(elemList.attr('value'));
+  const csrfToken = elemList.attr('value') as string; // guaranteed to be string by assertion
+
+  return { $, csrfToken };
+}
+
+/**
+ * Creates a new group in the given assessment as the user with the given CSRF token
+ */
+async function createGroup(
+  groupName: string,
+  csrfToken: string,
+  assessmentUrl: string,
+): Promise<cheerio.Root> {
+  const res = await fetch(assessmentUrl, {
+    method: 'POST',
+    body: new URLSearchParams({
+      __action: 'create_group',
+      __csrf_token: csrfToken,
+      groupName: groupName,
+    }),
+  });
+  assert.isOk(res.ok);
+  const $ = cheerio.load(await res.text());
+  return $;
+}
+
+/**
+ * Joins a group in an assessment using the provided join code as the user with the given CSRF token
+ */
+async function joinGroup(
+  assessmentUrl: string,
+  joinCode: string,
+  csrfToken: string,
+): Promise<cheerio.Root> {
+  const res = await fetch(assessmentUrl, {
+    method: 'POST',
+    body: new URLSearchParams({
+      __action: 'join_group',
+      __csrf_token: csrfToken,
+      join_code: joinCode,
+    }),
+  });
+  assert.isOk(res.ok);
+  const $ = cheerio.load(await res.text());
+  return $;
+}
+
+describe('Group based exam assessments', function () {
+  this.timeout(20000);
+
+  before('set up testing server', helperServer.before(TEST_COURSE_PATH));
+  before('set authenticated user', function (callback) {
+    storedConfig.authUid = config.authUid;
+    storedConfig.authName = config.authName;
+    storedConfig.authUin = config.authUin;
+    callback(null);
+  });
+
+  after('shut down testing server', helperServer.after);
+  after('unset authenticated user', function (callback) {
+    Object.assign(config, storedConfig);
+    callback(null);
+  });
+
+  describe('instructor access for exam assessment', function () {
+    step("should load the group tab for the first assessment's instructor URL", async function () {
+      // Get exam assessment URLs using ids from database
+      const result = await queryAsync(sql.select_group_work_exam_assessment, []);
+      assert.lengthOf(result.rows, 2);
+      assert.isDefined(result.rows[0].id);
+      const assessmentId = result.rows[0].id;
+      const instructorAssessmentsUrlGroupTab =
+        courseInstanceUrl + '/instructor/assessment/' + assessmentId + '/groups';
+
+      // Page should load successfully
+      const res = await fetch(instructorAssessmentsUrlGroupTab);
+      const page = await res.text();
+
+      // Check for correct CSRF tokens
+      const $ = cheerio.load(page);
+      const elemList = $('form input[name="__csrf_token"]');
+      assert.lengthOf(elemList, 5);
+      assert.nestedProperty(elemList[0], 'attribs.value');
+      const csrfToken = elemList.attr('value');
+      assert.isString(csrfToken);
```

Comment:
This appears to check for *any* CSRF tokens, and seems unnecessarily fragile (what if we add more forms to the page?). Was this check actually useful during development/testing? IMO we could likely remove it; it looks like `switchUserAndLoadAssessment` will find and return CSRF tokens that we then use in the response.

Reply:
Good point; it was a previous pattern in group assessment tests to check for the number of CSRF tokens every time a page was loaded, but I agree that it can easily break if more forms are added to the page.